{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create the dataset\n",
    "This notebook executes the refined code that was created in the `building_the_dataset.ipynb` notebook. This is to have in an organized way the code that worked, without eliminating the failed experiments. The code that didn't work in the beginning can give me ideas for future problems."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Steps to create the alternative sequences and dataset\n",
    "1. Load GWAS Catalog\n",
    "* Filter by IDs and keep the ones with the `rs` substring\n",
    "2. Load the chromosome's data\n",
    "* Filter the variants by their IDs if they appear in the GWAS Catalog, and by their type of sequence alteration (SNV)\n",
    "3. Generate sequences\n",
    "* Create the list of alleles for multi-allelic variants and the bed column\n",
    "* Generate the sequences with samtools based on the bed column\n",
    "4. Modify the sequences based on the alternative alleles list\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load modules and libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import MutableSeq, Seq\n",
    "import subprocess\n",
    "import numpy as np\n",
    "from numpy.random import seed\n",
    "import torch\n",
    "from torch.utils.data import Dataset, DataLoader"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Useful paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "databases_path = \"/mnt/sda1/Databases/\"\n",
    "# GWAS Catalog path\n",
    "gwas_catalog_path = os.path.join(databases_path, \"GWAS_Catalog_DATA/gwas_catalog_v1.0.2-associations_e110_r2023-09-25.tsv\")\n",
    "# Ensembl Variation path\n",
    "ensembl_path = os.path.join(databases_path, \"Ensembl/Variation/110/\")\n",
    "# Chromosomes' data path\n",
    "chromosomes_path = os.path.join(ensembl_path, \"chromosomes_data/\")\n",
    "# Reference genome path\n",
    "ref_genome_path = os.path.join(databases_path,\"Reference_Genome/GRCh38p14/Ensembl/Homo_sapiens_GRCh38_dna_primary_assembly.fa\")\n",
    "# GWAS Associated bed and sequences path\n",
    "gwas_associated_bed_path = os.path.join(databases_path, \"Ensembl/Variation/110/gwas_associated_sequences/beds\")\n",
    "gwas_associated_seq_path = os.path.join(databases_path, \"Ensembl/Variation/110/gwas_associated_sequences/ref_sequences\")\n",
    "rand_bed_path = os.path.join(databases_path, \"Ensembl/Variation/110/random_sequences/beds\")\n",
    "rand_seq_path = os.path.join(databases_path, \"Ensembl/Variation/110/random_sequences/ref_sequences\")\n",
    "# Datasets path\n",
    "dataset_path = os.path.join(ensembl_path, \"chromosome_datasets/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load GWAS Catalog"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATE ADDED TO CATALOG</th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>FIRST AUTHOR</th>\n",
       "      <th>DATE</th>\n",
       "      <th>JOURNAL</th>\n",
       "      <th>LINK</th>\n",
       "      <th>STUDY</th>\n",
       "      <th>DISEASE/TRAIT</th>\n",
       "      <th>INITIAL SAMPLE SIZE</th>\n",
       "      <th>REPLICATION SAMPLE SIZE</th>\n",
       "      <th>...</th>\n",
       "      <th>PVALUE_MLOG</th>\n",
       "      <th>P-VALUE (TEXT)</th>\n",
       "      <th>OR or BETA</th>\n",
       "      <th>95% CI (TEXT)</th>\n",
       "      <th>PLATFORM [SNPS PASSING QC]</th>\n",
       "      <th>CNV</th>\n",
       "      <th>MAPPED_TRAIT</th>\n",
       "      <th>MAPPED_TRAIT_URI</th>\n",
       "      <th>STUDY ACCESSION</th>\n",
       "      <th>GENOTYPING TECHNOLOGY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2008-07-01</td>\n",
       "      <td>18391951</td>\n",
       "      <td>Gudbjartsson DF</td>\n",
       "      <td>2008-04-06</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/18391951</td>\n",
       "      <td>Many sequence variants affecting diversity of ...</td>\n",
       "      <td>Height</td>\n",
       "      <td>30,968 European ancestry individuals</td>\n",
       "      <td>8,541 European ancestry individuals</td>\n",
       "      <td>...</td>\n",
       "      <td>27.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.4</td>\n",
       "      <td>[6.03-8.77] % s.d. increase</td>\n",
       "      <td>Affymetrix, Illumina [up to 304226]</td>\n",
       "      <td>N</td>\n",
       "      <td>body height</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004339</td>\n",
       "      <td>GCST000175</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2008-07-01</td>\n",
       "      <td>18391951</td>\n",
       "      <td>Gudbjartsson DF</td>\n",
       "      <td>2008-04-06</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/18391951</td>\n",
       "      <td>Many sequence variants affecting diversity of ...</td>\n",
       "      <td>Height</td>\n",
       "      <td>30,968 European ancestry individuals</td>\n",
       "      <td>8,541 European ancestry individuals</td>\n",
       "      <td>...</td>\n",
       "      <td>5.522878745280337</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.5</td>\n",
       "      <td>[1.93-5.07] % s.d. increase</td>\n",
       "      <td>Affymetrix, Illumina [up to 304226]</td>\n",
       "      <td>N</td>\n",
       "      <td>body height</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004339</td>\n",
       "      <td>GCST000175</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2008-07-01</td>\n",
       "      <td>18391951</td>\n",
       "      <td>Gudbjartsson DF</td>\n",
       "      <td>2008-04-06</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/18391951</td>\n",
       "      <td>Many sequence variants affecting diversity of ...</td>\n",
       "      <td>Height</td>\n",
       "      <td>30,968 European ancestry individuals</td>\n",
       "      <td>8,541 European ancestry individuals</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.6</td>\n",
       "      <td>[2.23-4.97] % s.d. increase</td>\n",
       "      <td>Affymetrix, Illumina [up to 304226]</td>\n",
       "      <td>N</td>\n",
       "      <td>body height</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004339</td>\n",
       "      <td>GCST000175</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATE ADDED TO CATALOG  PUBMEDID     FIRST AUTHOR        DATE    JOURNAL  \\\n",
       "0            2008-07-01  18391951  Gudbjartsson DF  2008-04-06  Nat Genet   \n",
       "1            2008-07-01  18391951  Gudbjartsson DF  2008-04-06  Nat Genet   \n",
       "2            2008-07-01  18391951  Gudbjartsson DF  2008-04-06  Nat Genet   \n",
       "\n",
       "                                   LINK  \\\n",
       "0  www.ncbi.nlm.nih.gov/pubmed/18391951   \n",
       "1  www.ncbi.nlm.nih.gov/pubmed/18391951   \n",
       "2  www.ncbi.nlm.nih.gov/pubmed/18391951   \n",
       "\n",
       "                                               STUDY DISEASE/TRAIT  \\\n",
       "0  Many sequence variants affecting diversity of ...        Height   \n",
       "1  Many sequence variants affecting diversity of ...        Height   \n",
       "2  Many sequence variants affecting diversity of ...        Height   \n",
       "\n",
       "                    INITIAL SAMPLE SIZE              REPLICATION SAMPLE SIZE  \\\n",
       "0  30,968 European ancestry individuals  8,541 European ancestry individuals   \n",
       "1  30,968 European ancestry individuals  8,541 European ancestry individuals   \n",
       "2  30,968 European ancestry individuals  8,541 European ancestry individuals   \n",
       "\n",
       "   ...        PVALUE_MLOG P-VALUE (TEXT) OR or BETA  \\\n",
       "0  ...               27.0            NaN        7.4   \n",
       "1  ...  5.522878745280337            NaN        3.5   \n",
       "2  ...                7.0            NaN        3.6   \n",
       "\n",
       "                 95% CI (TEXT)           PLATFORM [SNPS PASSING QC] CNV  \\\n",
       "0  [6.03-8.77] % s.d. increase  Affymetrix, Illumina [up to 304226]   N   \n",
       "1  [1.93-5.07] % s.d. increase  Affymetrix, Illumina [up to 304226]   N   \n",
       "2  [2.23-4.97] % s.d. increase  Affymetrix, Illumina [up to 304226]   N   \n",
       "\n",
       "  MAPPED_TRAIT                      MAPPED_TRAIT_URI STUDY ACCESSION  \\\n",
       "0  body height  http://www.ebi.ac.uk/efo/EFO_0004339      GCST000175   \n",
       "1  body height  http://www.ebi.ac.uk/efo/EFO_0004339      GCST000175   \n",
       "2  body height  http://www.ebi.ac.uk/efo/EFO_0004339      GCST000175   \n",
       "\n",
       "          GENOTYPING TECHNOLOGY  \n",
       "0  Genome-wide genotyping array  \n",
       "1  Genome-wide genotyping array  \n",
       "2  Genome-wide genotyping array  \n",
       "\n",
       "[3 rows x 38 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_catalog = pd.read_csv(gwas_catalog_path, delimiter=\"\\t\", dtype=str)\n",
    "gwas_catalog.head(n=3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter by IDs containing the `rs` substring"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATE ADDED TO CATALOG</th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>FIRST AUTHOR</th>\n",
       "      <th>DATE</th>\n",
       "      <th>JOURNAL</th>\n",
       "      <th>LINK</th>\n",
       "      <th>STUDY</th>\n",
       "      <th>DISEASE/TRAIT</th>\n",
       "      <th>INITIAL SAMPLE SIZE</th>\n",
       "      <th>REPLICATION SAMPLE SIZE</th>\n",
       "      <th>...</th>\n",
       "      <th>PVALUE_MLOG</th>\n",
       "      <th>P-VALUE (TEXT)</th>\n",
       "      <th>OR or BETA</th>\n",
       "      <th>95% CI (TEXT)</th>\n",
       "      <th>PLATFORM [SNPS PASSING QC]</th>\n",
       "      <th>CNV</th>\n",
       "      <th>MAPPED_TRAIT</th>\n",
       "      <th>MAPPED_TRAIT_URI</th>\n",
       "      <th>STUDY ACCESSION</th>\n",
       "      <th>GENOTYPING TECHNOLOGY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2008-07-01</td>\n",
       "      <td>18391951</td>\n",
       "      <td>Gudbjartsson DF</td>\n",
       "      <td>2008-04-06</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/18391951</td>\n",
       "      <td>Many sequence variants affecting diversity of ...</td>\n",
       "      <td>Height</td>\n",
       "      <td>30,968 European ancestry individuals</td>\n",
       "      <td>8,541 European ancestry individuals</td>\n",
       "      <td>...</td>\n",
       "      <td>27.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.4</td>\n",
       "      <td>[6.03-8.77] % s.d. increase</td>\n",
       "      <td>Affymetrix, Illumina [up to 304226]</td>\n",
       "      <td>N</td>\n",
       "      <td>body height</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004339</td>\n",
       "      <td>GCST000175</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2008-07-01</td>\n",
       "      <td>18391951</td>\n",
       "      <td>Gudbjartsson DF</td>\n",
       "      <td>2008-04-06</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/18391951</td>\n",
       "      <td>Many sequence variants affecting diversity of ...</td>\n",
       "      <td>Height</td>\n",
       "      <td>30,968 European ancestry individuals</td>\n",
       "      <td>8,541 European ancestry individuals</td>\n",
       "      <td>...</td>\n",
       "      <td>5.522878745280337</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.5</td>\n",
       "      <td>[1.93-5.07] % s.d. increase</td>\n",
       "      <td>Affymetrix, Illumina [up to 304226]</td>\n",
       "      <td>N</td>\n",
       "      <td>body height</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004339</td>\n",
       "      <td>GCST000175</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2008-07-01</td>\n",
       "      <td>18391951</td>\n",
       "      <td>Gudbjartsson DF</td>\n",
       "      <td>2008-04-06</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/18391951</td>\n",
       "      <td>Many sequence variants affecting diversity of ...</td>\n",
       "      <td>Height</td>\n",
       "      <td>30,968 European ancestry individuals</td>\n",
       "      <td>8,541 European ancestry individuals</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.6</td>\n",
       "      <td>[2.23-4.97] % s.d. increase</td>\n",
       "      <td>Affymetrix, Illumina [up to 304226]</td>\n",
       "      <td>N</td>\n",
       "      <td>body height</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004339</td>\n",
       "      <td>GCST000175</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATE ADDED TO CATALOG  PUBMEDID     FIRST AUTHOR        DATE    JOURNAL  \\\n",
       "0            2008-07-01  18391951  Gudbjartsson DF  2008-04-06  Nat Genet   \n",
       "1            2008-07-01  18391951  Gudbjartsson DF  2008-04-06  Nat Genet   \n",
       "2            2008-07-01  18391951  Gudbjartsson DF  2008-04-06  Nat Genet   \n",
       "\n",
       "                                   LINK  \\\n",
       "0  www.ncbi.nlm.nih.gov/pubmed/18391951   \n",
       "1  www.ncbi.nlm.nih.gov/pubmed/18391951   \n",
       "2  www.ncbi.nlm.nih.gov/pubmed/18391951   \n",
       "\n",
       "                                               STUDY DISEASE/TRAIT  \\\n",
       "0  Many sequence variants affecting diversity of ...        Height   \n",
       "1  Many sequence variants affecting diversity of ...        Height   \n",
       "2  Many sequence variants affecting diversity of ...        Height   \n",
       "\n",
       "                    INITIAL SAMPLE SIZE              REPLICATION SAMPLE SIZE  \\\n",
       "0  30,968 European ancestry individuals  8,541 European ancestry individuals   \n",
       "1  30,968 European ancestry individuals  8,541 European ancestry individuals   \n",
       "2  30,968 European ancestry individuals  8,541 European ancestry individuals   \n",
       "\n",
       "   ...        PVALUE_MLOG P-VALUE (TEXT) OR or BETA  \\\n",
       "0  ...               27.0            NaN        7.4   \n",
       "1  ...  5.522878745280337            NaN        3.5   \n",
       "2  ...                7.0            NaN        3.6   \n",
       "\n",
       "                 95% CI (TEXT)           PLATFORM [SNPS PASSING QC] CNV  \\\n",
       "0  [6.03-8.77] % s.d. increase  Affymetrix, Illumina [up to 304226]   N   \n",
       "1  [1.93-5.07] % s.d. increase  Affymetrix, Illumina [up to 304226]   N   \n",
       "2  [2.23-4.97] % s.d. increase  Affymetrix, Illumina [up to 304226]   N   \n",
       "\n",
       "  MAPPED_TRAIT                      MAPPED_TRAIT_URI STUDY ACCESSION  \\\n",
       "0  body height  http://www.ebi.ac.uk/efo/EFO_0004339      GCST000175   \n",
       "1  body height  http://www.ebi.ac.uk/efo/EFO_0004339      GCST000175   \n",
       "2  body height  http://www.ebi.ac.uk/efo/EFO_0004339      GCST000175   \n",
       "\n",
       "          GENOTYPING TECHNOLOGY  \n",
       "0  Genome-wide genotyping array  \n",
       "1  Genome-wide genotyping array  \n",
       "2  Genome-wide genotyping array  \n",
       "\n",
       "[3 rows x 38 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract all the variants containing the `rs` substring and creating a new Series object to not affect the original dataframe\n",
    "gwas_catalog_rs_filtered = gwas_catalog[gwas_catalog[\"SNPS\"].str.contains(\"rs\", case=False, na=False)].copy(deep=True)\n",
    "gwas_catalog_rs_filtered.head(n=3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load chromosome's data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>tsa</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>10007</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs1572047073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>10014</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>insertion</td>\n",
       "      <td>rs1558169263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>10017</td>\n",
       "      <td>CACCC</td>\n",
       "      <td>C</td>\n",
       "      <td>indel</td>\n",
       "      <td>rs1558169385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>10018</td>\n",
       "      <td>A</td>\n",
       "      <td>AACA,AACG</td>\n",
       "      <td>insertion</td>\n",
       "      <td>rs1558169388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>10019</td>\n",
       "      <td>CC</td>\n",
       "      <td>C</td>\n",
       "      <td>indel</td>\n",
       "      <td>rs1558169386</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr    pos    ref        alt        tsa            id\n",
       "0   2  10007      C          A        SNV  rs1572047073\n",
       "1   2  10014      C         CG  insertion  rs1558169263\n",
       "2   2  10017  CACCC          C      indel  rs1558169385\n",
       "3   2  10018      A  AACA,AACG  insertion  rs1558169388\n",
       "4   2  10019     CC          C      indel  rs1558169386"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chromosome = \"2\"\n",
    "chr2_data = pd.read_csv(os.path.join(chromosomes_path, \"chr{}_data.tsv\".format(chromosome)), delimiter=\"\\t\", dtype=str)\n",
    "# Change the column names for easier access when referring to the dataframe\n",
    "chr2_data.rename(columns={'#[1]CHROM':'chr', '[2]POS':'pos', '[3]REF':'ref', '[4]ALT':'alt', '[5]TSA':'tsa', '[6]ID':'id'}, inplace=True)\n",
    "chr2_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter variants which IDs are in the GWAS Catalog and by type of sequence alteration (SNV)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>tsa</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9074</th>\n",
       "      <td>2</td>\n",
       "      <td>38938</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs11542478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14066</th>\n",
       "      <td>2</td>\n",
       "      <td>58639</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs62116661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18575</th>\n",
       "      <td>2</td>\n",
       "      <td>76417</td>\n",
       "      <td>T</td>\n",
       "      <td>A,C,G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs300769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18619</th>\n",
       "      <td>2</td>\n",
       "      <td>76530</td>\n",
       "      <td>C</td>\n",
       "      <td>G,T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs300768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21715</th>\n",
       "      <td>2</td>\n",
       "      <td>89910</td>\n",
       "      <td>G</td>\n",
       "      <td>A,C,T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs300789</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr    pos ref    alt  tsa          id\n",
       "9074    2  38938   A      C  SNV  rs11542478\n",
       "14066   2  58639   C      T  SNV  rs62116661\n",
       "18575   2  76417   T  A,C,G  SNV    rs300769\n",
       "18619   2  76530   C    G,T  SNV    rs300768\n",
       "21715   2  89910   G  A,C,T  SNV    rs300789"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract the gwas associated variants in chromosome 2 into a new data frame\n",
    "gwas_catalog_chr2 = gwas_catalog_rs_filtered[gwas_catalog_rs_filtered[\"CHR_ID\"]==\"2\"]\n",
    "# Filter the chromosome's variants by the ids registered in the gwas catalog and by the type of sequence alteration (SNV)\n",
    "chr2_gwas_snps = chr2_data[(chr2_data[\"id\"].isin(gwas_catalog_chr2.SNPS)) & (chr2_data[\"tsa\"]==\"SNV\") ].copy(deep=True)\n",
    "chr2_gwas_snps.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create the column containing the alleles and the bed-like column\n",
    "Create the list of possible alleles if the variant is multi-allelic, and the bed strings to extract the subsequences from the reference genome with `samtools`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>tsa</th>\n",
       "      <th>id</th>\n",
       "      <th>alt_list</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>bed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9074</th>\n",
       "      <td>2</td>\n",
       "      <td>38938</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs11542478</td>\n",
       "      <td>[C]</td>\n",
       "      <td>38875</td>\n",
       "      <td>39002</td>\n",
       "      <td>2:38875-39002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14066</th>\n",
       "      <td>2</td>\n",
       "      <td>58639</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs62116661</td>\n",
       "      <td>[T]</td>\n",
       "      <td>58576</td>\n",
       "      <td>58703</td>\n",
       "      <td>2:58576-58703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18575</th>\n",
       "      <td>2</td>\n",
       "      <td>76417</td>\n",
       "      <td>T</td>\n",
       "      <td>A,C,G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs300769</td>\n",
       "      <td>[A, C, G]</td>\n",
       "      <td>76354</td>\n",
       "      <td>76481</td>\n",
       "      <td>2:76354-76481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18619</th>\n",
       "      <td>2</td>\n",
       "      <td>76530</td>\n",
       "      <td>C</td>\n",
       "      <td>G,T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs300768</td>\n",
       "      <td>[G, T]</td>\n",
       "      <td>76467</td>\n",
       "      <td>76594</td>\n",
       "      <td>2:76467-76594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21715</th>\n",
       "      <td>2</td>\n",
       "      <td>89910</td>\n",
       "      <td>G</td>\n",
       "      <td>A,C,T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs300789</td>\n",
       "      <td>[A, C, T]</td>\n",
       "      <td>89847</td>\n",
       "      <td>89974</td>\n",
       "      <td>2:89847-89974</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr    pos ref    alt  tsa          id   alt_list  start    end  \\\n",
       "9074    2  38938   A      C  SNV  rs11542478        [C]  38875  39002   \n",
       "14066   2  58639   C      T  SNV  rs62116661        [T]  58576  58703   \n",
       "18575   2  76417   T  A,C,G  SNV    rs300769  [A, C, G]  76354  76481   \n",
       "18619   2  76530   C    G,T  SNV    rs300768     [G, T]  76467  76594   \n",
       "21715   2  89910   G  A,C,T  SNV    rs300789  [A, C, T]  89847  89974   \n",
       "\n",
       "                 bed  \n",
       "9074   2:38875-39002  \n",
       "14066  2:58576-58703  \n",
       "18575  2:76354-76481  \n",
       "18619  2:76467-76594  \n",
       "21715  2:89847-89974  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create the list of alternative alleles\n",
    "chr2_gwas_snps[\"alt_list\"] = chr2_gwas_snps[\"alt\"].str.split(pat=\",\")\n",
    "# Create the bed column\n",
    "chr2_gwas_snps['start'] = chr2_gwas_snps['pos'].astype(int) - 63\n",
    "chr2_gwas_snps['end'] = chr2_gwas_snps['pos'].astype(int) + 64\n",
    "chr2_gwas_snps['bed'] = chr2_gwas_snps['chr'].astype(str) + ':' + chr2_gwas_snps['start'].astype(str) + '-' + chr2_gwas_snps['end'].astype(str)\n",
    "chr2_gwas_snps.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(21232, 10)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr2_gwas_snps.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate the sequences with samtools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "exit_code = subprocess.Popen(\"samtools faidx {} -r {}/bed_chr{} -o {}/ref_seq_chr{}\".format(ref_genome_path, gwas_associated_bed_path, chromosome,\n",
    "                                                                                            gwas_associated_seq_path, chromosome), \n",
    "                                    shell=True, stdout=subprocess.PIPE).stdout.read()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load the created sequences and add them to the dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>tsa</th>\n",
       "      <th>id</th>\n",
       "      <th>alt_list</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>bed</th>\n",
       "      <th>ref_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9074</th>\n",
       "      <td>2</td>\n",
       "      <td>38938</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs11542478</td>\n",
       "      <td>[C]</td>\n",
       "      <td>38875</td>\n",
       "      <td>39002</td>\n",
       "      <td>2:38875-39002</td>\n",
       "      <td>TGCAAATAGTGTATAGAAAAAGCTCTGTTTAGAAACTGCCATAGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14066</th>\n",
       "      <td>2</td>\n",
       "      <td>58639</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs62116661</td>\n",
       "      <td>[T]</td>\n",
       "      <td>58576</td>\n",
       "      <td>58703</td>\n",
       "      <td>2:58576-58703</td>\n",
       "      <td>CTCCACAATAGTCAAAATGAAAGAAAAATACCAAGCCTCTCTCAGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18575</th>\n",
       "      <td>2</td>\n",
       "      <td>76417</td>\n",
       "      <td>T</td>\n",
       "      <td>A,C,G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs300769</td>\n",
       "      <td>[A, C, G]</td>\n",
       "      <td>76354</td>\n",
       "      <td>76481</td>\n",
       "      <td>2:76354-76481</td>\n",
       "      <td>TTCTCACACTGCCAATAAAAACATAGCTAAGACTGGGTAATTTATA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18619</th>\n",
       "      <td>2</td>\n",
       "      <td>76530</td>\n",
       "      <td>C</td>\n",
       "      <td>G,T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs300768</td>\n",
       "      <td>[G, T]</td>\n",
       "      <td>76467</td>\n",
       "      <td>76594</td>\n",
       "      <td>2:76467-76594</td>\n",
       "      <td>GGCAAAGGAGGAGCAAAGGCACCTCTTACATGGTGGCAGGCAAGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21715</th>\n",
       "      <td>2</td>\n",
       "      <td>89910</td>\n",
       "      <td>G</td>\n",
       "      <td>A,C,T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs300789</td>\n",
       "      <td>[A, C, T]</td>\n",
       "      <td>89847</td>\n",
       "      <td>89974</td>\n",
       "      <td>2:89847-89974</td>\n",
       "      <td>AGTACAGGACACACCATCGCAATAAATTAAAAAGGCAAATGCAAAT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr    pos ref    alt  tsa          id   alt_list  start    end  \\\n",
       "9074    2  38938   A      C  SNV  rs11542478        [C]  38875  39002   \n",
       "14066   2  58639   C      T  SNV  rs62116661        [T]  58576  58703   \n",
       "18575   2  76417   T  A,C,G  SNV    rs300769  [A, C, G]  76354  76481   \n",
       "18619   2  76530   C    G,T  SNV    rs300768     [G, T]  76467  76594   \n",
       "21715   2  89910   G  A,C,T  SNV    rs300789  [A, C, T]  89847  89974   \n",
       "\n",
       "                 bed                                            ref_seq  \n",
       "9074   2:38875-39002  TGCAAATAGTGTATAGAAAAAGCTCTGTTTAGAAACTGCCATAGCA...  \n",
       "14066  2:58576-58703  CTCCACAATAGTCAAAATGAAAGAAAAATACCAAGCCTCTCTCAGC...  \n",
       "18575  2:76354-76481  TTCTCACACTGCCAATAAAAACATAGCTAAGACTGGGTAATTTATA...  \n",
       "18619  2:76467-76594  GGCAAAGGAGGAGCAAAGGCACCTCTTACATGGTGGCAGGCAAGAG...  \n",
       "21715  2:89847-89974  AGTACAGGACACACCATCGCAATAAATTAAAAAGGCAAATGCAAAT...  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "records = list(SeqIO.parse(gwas_associated_seq_path+'ref_seq_chr{}'.format(chromosome), \"fasta\"))\n",
    "ref_seqs = [str(sequence[1].seq) for sequence in enumerate(records)]\n",
    "chr2_gwas_snps[\"ref_seq\"] = ref_seqs\n",
    "chr2_gwas_snps.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Modify the sequences based on the information of each variant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>tsa</th>\n",
       "      <th>id</th>\n",
       "      <th>alt_list</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>bed</th>\n",
       "      <th>ref_seq</th>\n",
       "      <th>alt_seqs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9074</th>\n",
       "      <td>2</td>\n",
       "      <td>38938</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs11542478</td>\n",
       "      <td>[C]</td>\n",
       "      <td>38875</td>\n",
       "      <td>39002</td>\n",
       "      <td>2:38875-39002</td>\n",
       "      <td>TGCAAATAGTGTATAGAAAAAGCTCTGTTTAGAAACTGCCATAGCA...</td>\n",
       "      <td>[TGCAAATAGTGTATAGAAAAAGCTCTGTTTAGAAACTGCCATAGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14066</th>\n",
       "      <td>2</td>\n",
       "      <td>58639</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs62116661</td>\n",
       "      <td>[T]</td>\n",
       "      <td>58576</td>\n",
       "      <td>58703</td>\n",
       "      <td>2:58576-58703</td>\n",
       "      <td>CTCCACAATAGTCAAAATGAAAGAAAAATACCAAGCCTCTCTCAGC...</td>\n",
       "      <td>[CTCCACAATAGTCAAAATGAAAGAAAAATACCAAGCCTCTCTCAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18575</th>\n",
       "      <td>2</td>\n",
       "      <td>76417</td>\n",
       "      <td>T</td>\n",
       "      <td>A,C,G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs300769</td>\n",
       "      <td>[A, C, G]</td>\n",
       "      <td>76354</td>\n",
       "      <td>76481</td>\n",
       "      <td>2:76354-76481</td>\n",
       "      <td>TTCTCACACTGCCAATAAAAACATAGCTAAGACTGGGTAATTTATA...</td>\n",
       "      <td>[TTCTCACACTGCCAATAAAAACATAGCTAAGACTGGGTAATTTAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18619</th>\n",
       "      <td>2</td>\n",
       "      <td>76530</td>\n",
       "      <td>C</td>\n",
       "      <td>G,T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs300768</td>\n",
       "      <td>[G, T]</td>\n",
       "      <td>76467</td>\n",
       "      <td>76594</td>\n",
       "      <td>2:76467-76594</td>\n",
       "      <td>GGCAAAGGAGGAGCAAAGGCACCTCTTACATGGTGGCAGGCAAGAG...</td>\n",
       "      <td>[GGCAAAGGAGGAGCAAAGGCACCTCTTACATGGTGGCAGGCAAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21715</th>\n",
       "      <td>2</td>\n",
       "      <td>89910</td>\n",
       "      <td>G</td>\n",
       "      <td>A,C,T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs300789</td>\n",
       "      <td>[A, C, T]</td>\n",
       "      <td>89847</td>\n",
       "      <td>89974</td>\n",
       "      <td>2:89847-89974</td>\n",
       "      <td>AGTACAGGACACACCATCGCAATAAATTAAAAAGGCAAATGCAAAT...</td>\n",
       "      <td>[AGTACAGGACACACCATCGCAATAAATTAAAAAGGCAAATGCAAA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr    pos ref    alt  tsa          id   alt_list  start    end  \\\n",
       "9074    2  38938   A      C  SNV  rs11542478        [C]  38875  39002   \n",
       "14066   2  58639   C      T  SNV  rs62116661        [T]  58576  58703   \n",
       "18575   2  76417   T  A,C,G  SNV    rs300769  [A, C, G]  76354  76481   \n",
       "18619   2  76530   C    G,T  SNV    rs300768     [G, T]  76467  76594   \n",
       "21715   2  89910   G  A,C,T  SNV    rs300789  [A, C, T]  89847  89974   \n",
       "\n",
       "                 bed                                            ref_seq  \\\n",
       "9074   2:38875-39002  TGCAAATAGTGTATAGAAAAAGCTCTGTTTAGAAACTGCCATAGCA...   \n",
       "14066  2:58576-58703  CTCCACAATAGTCAAAATGAAAGAAAAATACCAAGCCTCTCTCAGC...   \n",
       "18575  2:76354-76481  TTCTCACACTGCCAATAAAAACATAGCTAAGACTGGGTAATTTATA...   \n",
       "18619  2:76467-76594  GGCAAAGGAGGAGCAAAGGCACCTCTTACATGGTGGCAGGCAAGAG...   \n",
       "21715  2:89847-89974  AGTACAGGACACACCATCGCAATAAATTAAAAAGGCAAATGCAAAT...   \n",
       "\n",
       "                                                alt_seqs  \n",
       "9074   [TGCAAATAGTGTATAGAAAAAGCTCTGTTTAGAAACTGCCATAGC...  \n",
       "14066  [CTCCACAATAGTCAAAATGAAAGAAAAATACCAAGCCTCTCTCAG...  \n",
       "18575  [TTCTCACACTGCCAATAAAAACATAGCTAAGACTGGGTAATTTAT...  \n",
       "18619  [GGCAAAGGAGGAGCAAAGGCACCTCTTACATGGTGGCAGGCAAGA...  \n",
       "21715  [AGTACAGGACACACCATCGCAATAAATTAAAAAGGCAAATGCAAA...  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Generate the subsequences for variants, including the multiallelic ones\n",
    "alt_seq = []\n",
    "for idx, variant in enumerate(chr2_gwas_snps['alt_list']):\n",
    "    tmp_alt_seq = [] # Clear the contents of this list each time the for loop goes to a new register\n",
    "    for allele in variant:\n",
    "        tmp_seq = MutableSeq(chr2_gwas_snps['ref_seq'].iloc[idx])\n",
    "        tmp_seq[63] = allele\n",
    "        tmp_seq = str(tmp_seq)\n",
    "        tmp_alt_seq.append(tmp_seq)\n",
    "    alt_seq.append(tmp_alt_seq)\n",
    "\n",
    "chr2_gwas_snps[\"alt_seqs\"] = alt_seq\n",
    "chr2_gwas_snps.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save the dataframe\n",
    "This data frame contains the reference and alternative sequences of the GWAS associated variants in chromosome 2:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "#chr2_gwas_snps.to_csv(os.path.join(dataset_path, \"chr2_gwas_dataset.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "21232"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(chr2_gwas_snps)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Eliminate the gwas-associated registers from the ensembl df, pick random samples and create their sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>tsa</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>10007</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs1572047073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>10014</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>insertion</td>\n",
       "      <td>rs1558169263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>10017</td>\n",
       "      <td>CACCC</td>\n",
       "      <td>C</td>\n",
       "      <td>indel</td>\n",
       "      <td>rs1558169385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>10018</td>\n",
       "      <td>A</td>\n",
       "      <td>AACA,AACG</td>\n",
       "      <td>insertion</td>\n",
       "      <td>rs1558169388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>10019</td>\n",
       "      <td>CC</td>\n",
       "      <td>C</td>\n",
       "      <td>indel</td>\n",
       "      <td>rs1558169386</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr    pos    ref        alt        tsa            id\n",
       "0   2  10007      C          A        SNV  rs1572047073\n",
       "1   2  10014      C         CG  insertion  rs1558169263\n",
       "2   2  10017  CACCC          C      indel  rs1558169385\n",
       "3   2  10018      A  AACA,AACG  insertion  rs1558169388\n",
       "4   2  10019     CC          C      indel  rs1558169386"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr2_data.drop(index=chr2_gwas_snps.index, inplace=True)\n",
    "chr2_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "chr               2\n",
       "pos           58647\n",
       "ref               C\n",
       "alt               G\n",
       "tsa             SNV\n",
       "id     rs1572070720\n",
       "Name: 14068, dtype: object"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The gwas-associated registers were removed from the orignal dataset, and by indexing the remaining registers different one appear.\n",
    "chr2_data.iloc[chr2_gwas_snps.index[1]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter the remaining chromosome 2 variants by their type and leave only the SNVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>tsa</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>10007</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs1572047073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2</td>\n",
       "      <td>10020</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs1572047087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2</td>\n",
       "      <td>10022</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs1572047090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2</td>\n",
       "      <td>10026</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs1366167113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2</td>\n",
       "      <td>10027</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs1572047092</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr    pos ref alt  tsa            id\n",
       "0    2  10007   C   A  SNV  rs1572047073\n",
       "7    2  10020   C   A  SNV  rs1572047087\n",
       "8    2  10022   A   C  SNV  rs1572047090\n",
       "9    2  10026   A   G  SNV  rs1366167113\n",
       "14   2  10027   C   A  SNV  rs1572047092"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr2_snps = chr2_data[chr2_data['tsa']=='SNV']\n",
    "chr2_snps.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>tsa</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2641070</th>\n",
       "      <td>2</td>\n",
       "      <td>10483775</td>\n",
       "      <td>C</td>\n",
       "      <td>A,T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs1329558856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51700147</th>\n",
       "      <td>2</td>\n",
       "      <td>213255429</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs1449218878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27453276</th>\n",
       "      <td>2</td>\n",
       "      <td>109750792</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs1438538645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55793546</th>\n",
       "      <td>2</td>\n",
       "      <td>230397595</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs1246211497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15278237</th>\n",
       "      <td>2</td>\n",
       "      <td>58860529</td>\n",
       "      <td>G</td>\n",
       "      <td>A,T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs768583468</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr        pos ref  alt  tsa            id\n",
       "2641070    2   10483775   C  A,T  SNV  rs1329558856\n",
       "51700147   2  213255429   G    A  SNV  rs1449218878\n",
       "27453276   2  109750792   C    T  SNV  rs1438538645\n",
       "55793546   2  230397595   C    T  SNV  rs1246211497\n",
       "15278237   2   58860529   G  A,T  SNV   rs768583468"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Generate a seed for reproducibility\n",
    "seed = seed(20231122)\n",
    "chr2_rand_samples = chr2_snps.sample(n=len(chr2_gwas_snps), random_state=seed).copy(deep=True)\n",
    "chr2_rand_samples.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "from process_data import generate_bed, generate_sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>tsa</th>\n",
       "      <th>id</th>\n",
       "      <th>alt_list</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>bed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2641070</th>\n",
       "      <td>2</td>\n",
       "      <td>10483775</td>\n",
       "      <td>C</td>\n",
       "      <td>A,T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs1329558856</td>\n",
       "      <td>[A, T]</td>\n",
       "      <td>10483712</td>\n",
       "      <td>10483839</td>\n",
       "      <td>2:10483712-10483839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51700147</th>\n",
       "      <td>2</td>\n",
       "      <td>213255429</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs1449218878</td>\n",
       "      <td>[A]</td>\n",
       "      <td>213255366</td>\n",
       "      <td>213255493</td>\n",
       "      <td>2:213255366-213255493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27453276</th>\n",
       "      <td>2</td>\n",
       "      <td>109750792</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs1438538645</td>\n",
       "      <td>[T]</td>\n",
       "      <td>109750729</td>\n",
       "      <td>109750856</td>\n",
       "      <td>2:109750729-109750856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55793546</th>\n",
       "      <td>2</td>\n",
       "      <td>230397595</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs1246211497</td>\n",
       "      <td>[T]</td>\n",
       "      <td>230397532</td>\n",
       "      <td>230397659</td>\n",
       "      <td>2:230397532-230397659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15278237</th>\n",
       "      <td>2</td>\n",
       "      <td>58860529</td>\n",
       "      <td>G</td>\n",
       "      <td>A,T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs768583468</td>\n",
       "      <td>[A, T]</td>\n",
       "      <td>58860466</td>\n",
       "      <td>58860593</td>\n",
       "      <td>2:58860466-58860593</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr        pos ref  alt  tsa            id alt_list      start  \\\n",
       "2641070    2   10483775   C  A,T  SNV  rs1329558856   [A, T]   10483712   \n",
       "51700147   2  213255429   G    A  SNV  rs1449218878      [A]  213255366   \n",
       "27453276   2  109750792   C    T  SNV  rs1438538645      [T]  109750729   \n",
       "55793546   2  230397595   C    T  SNV  rs1246211497      [T]  230397532   \n",
       "15278237   2   58860529   G  A,T  SNV   rs768583468   [A, T]   58860466   \n",
       "\n",
       "                end                    bed  \n",
       "2641070    10483839    2:10483712-10483839  \n",
       "51700147  213255493  2:213255366-213255493  \n",
       "27453276  109750856  2:109750729-109750856  \n",
       "55793546  230397659  2:230397532-230397659  \n",
       "15278237   58860593    2:58860466-58860593  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr2_rand_samples = generate_bed(chr2_rand_samples, gen_bed=True, res_path=rand_bed_path, chromosome=chromosome)\n",
    "chr2_rand_samples.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>tsa</th>\n",
       "      <th>id</th>\n",
       "      <th>alt_list</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>bed</th>\n",
       "      <th>ref_seq</th>\n",
       "      <th>alt_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2641070</th>\n",
       "      <td>2</td>\n",
       "      <td>10483775</td>\n",
       "      <td>C</td>\n",
       "      <td>A,T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs1329558856</td>\n",
       "      <td>[A, T]</td>\n",
       "      <td>10483712</td>\n",
       "      <td>10483839</td>\n",
       "      <td>2:10483712-10483839</td>\n",
       "      <td>TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTGT...</td>\n",
       "      <td>[TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51700147</th>\n",
       "      <td>2</td>\n",
       "      <td>213255429</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs1449218878</td>\n",
       "      <td>[A]</td>\n",
       "      <td>213255366</td>\n",
       "      <td>213255493</td>\n",
       "      <td>2:213255366-213255493</td>\n",
       "      <td>AGTATTATTTCTATTCTTTACTTGAGAATCCAGTTTTGTAGGACTT...</td>\n",
       "      <td>[AGTATTATTTCTATTCTTTACTTGAGAATCCAGTTTTGTAGGACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27453276</th>\n",
       "      <td>2</td>\n",
       "      <td>109750792</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs1438538645</td>\n",
       "      <td>[T]</td>\n",
       "      <td>109750729</td>\n",
       "      <td>109750856</td>\n",
       "      <td>2:109750729-109750856</td>\n",
       "      <td>ATTTTTTTTTTTCGAGTCAGAGTCTTGTTCTGTCACCCAGGCTGGA...</td>\n",
       "      <td>[ATTTTTTTTTTTCGAGTCAGAGTCTTGTTCTGTCACCCAGGCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55793546</th>\n",
       "      <td>2</td>\n",
       "      <td>230397595</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs1246211497</td>\n",
       "      <td>[T]</td>\n",
       "      <td>230397532</td>\n",
       "      <td>230397659</td>\n",
       "      <td>2:230397532-230397659</td>\n",
       "      <td>CTGCTTTGGGTTCTGCTGAAATCATGGATGAGTTCTTTCTTTAAAT...</td>\n",
       "      <td>[CTGCTTTGGGTTCTGCTGAAATCATGGATGAGTTCTTTCTTTAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15278237</th>\n",
       "      <td>2</td>\n",
       "      <td>58860529</td>\n",
       "      <td>G</td>\n",
       "      <td>A,T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs768583468</td>\n",
       "      <td>[A, T]</td>\n",
       "      <td>58860466</td>\n",
       "      <td>58860593</td>\n",
       "      <td>2:58860466-58860593</td>\n",
       "      <td>ATTAATATTTACCCTGTATATTTTCACAGGACCATTATATTGATCA...</td>\n",
       "      <td>[ATTAATATTTACCCTGTATATTTTCACAGGACCATTATATTGATC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr        pos ref  alt  tsa            id alt_list      start  \\\n",
       "2641070    2   10483775   C  A,T  SNV  rs1329558856   [A, T]   10483712   \n",
       "51700147   2  213255429   G    A  SNV  rs1449218878      [A]  213255366   \n",
       "27453276   2  109750792   C    T  SNV  rs1438538645      [T]  109750729   \n",
       "55793546   2  230397595   C    T  SNV  rs1246211497      [T]  230397532   \n",
       "15278237   2   58860529   G  A,T  SNV   rs768583468   [A, T]   58860466   \n",
       "\n",
       "                end                    bed  \\\n",
       "2641070    10483839    2:10483712-10483839   \n",
       "51700147  213255493  2:213255366-213255493   \n",
       "27453276  109750856  2:109750729-109750856   \n",
       "55793546  230397659  2:230397532-230397659   \n",
       "15278237   58860593    2:58860466-58860593   \n",
       "\n",
       "                                                    ref_seq  \\\n",
       "2641070   TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTGT...   \n",
       "51700147  AGTATTATTTCTATTCTTTACTTGAGAATCCAGTTTTGTAGGACTT...   \n",
       "27453276  ATTTTTTTTTTTCGAGTCAGAGTCTTGTTCTGTCACCCAGGCTGGA...   \n",
       "55793546  CTGCTTTGGGTTCTGCTGAAATCATGGATGAGTTCTTTCTTTAAAT...   \n",
       "15278237  ATTAATATTTACCCTGTATATTTTCACAGGACCATTATATTGATCA...   \n",
       "\n",
       "                                                    alt_seq  \n",
       "2641070   [TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTG...  \n",
       "51700147  [AGTATTATTTCTATTCTTTACTTGAGAATCCAGTTTTGTAGGACT...  \n",
       "27453276  [ATTTTTTTTTTTCGAGTCAGAGTCTTGTTCTGTCACCCAGGCTGG...  \n",
       "55793546  [CTGCTTTGGGTTCTGCTGAAATCATGGATGAGTTCTTTCTTTAAA...  \n",
       "15278237  [ATTAATATTTACCCTGTATATTTTCACAGGACCATTATATTGATC...  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr2_rand_samples = generate_sequences(variant_df=chr2_rand_samples, chromosome=chromosome, ref_genome_path=ref_genome_path,\n",
    "                                       seq_path=rand_seq_path, generate_fasta=True, bed_path=rand_bed_path)\n",
    "chr2_rand_samples.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTGTACTTTATTCAACCCTCTCGTTGATAGGTATTCACTTTGTTCCCACAGTTTTCCAGGGCTACAATCCTTGAATATCAAGCCTT \n",
      " TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTGTACTTTATTCAACCCTCTAGTTGATAGGTATTCACTTTGTTCCCACAGTTTTCCAGGGCTACAATCCTTGAATATCAAGCCTT \n",
      " TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTGTACTTTATTCAACCCTCTTGTTGATAGGTATTCACTTTGTTCCCACAGTTTTCCAGGGCTACAATCCTTGAATATCAAGCCTT\n"
     ]
    }
   ],
   "source": [
    "print('',chr2_rand_samples['ref_seq'].iloc[0],'\\n', chr2_rand_samples['alt_seq'].iloc[0][0],'\\n', \n",
    "      chr2_rand_samples['alt_seq'].iloc[0][1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save the random sequences in a csv file for easy access"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Commented because it makes no sense to run it multiple times\n",
    "#chr2_rand_samples.to_csv(os.path.join(dataset_path, 'chr2_rand_dataset.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Before the DataLoader"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Take a look at what we already have"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>tsa</th>\n",
       "      <th>id</th>\n",
       "      <th>alt_list</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>bed</th>\n",
       "      <th>ref_seq</th>\n",
       "      <th>alt_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2641070</th>\n",
       "      <td>2</td>\n",
       "      <td>10483775</td>\n",
       "      <td>C</td>\n",
       "      <td>A,T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs1329558856</td>\n",
       "      <td>[A, T]</td>\n",
       "      <td>10483712</td>\n",
       "      <td>10483839</td>\n",
       "      <td>2:10483712-10483839</td>\n",
       "      <td>TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTGT...</td>\n",
       "      <td>[TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51700147</th>\n",
       "      <td>2</td>\n",
       "      <td>213255429</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs1449218878</td>\n",
       "      <td>[A]</td>\n",
       "      <td>213255366</td>\n",
       "      <td>213255493</td>\n",
       "      <td>2:213255366-213255493</td>\n",
       "      <td>AGTATTATTTCTATTCTTTACTTGAGAATCCAGTTTTGTAGGACTT...</td>\n",
       "      <td>[AGTATTATTTCTATTCTTTACTTGAGAATCCAGTTTTGTAGGACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27453276</th>\n",
       "      <td>2</td>\n",
       "      <td>109750792</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs1438538645</td>\n",
       "      <td>[T]</td>\n",
       "      <td>109750729</td>\n",
       "      <td>109750856</td>\n",
       "      <td>2:109750729-109750856</td>\n",
       "      <td>ATTTTTTTTTTTCGAGTCAGAGTCTTGTTCTGTCACCCAGGCTGGA...</td>\n",
       "      <td>[ATTTTTTTTTTTCGAGTCAGAGTCTTGTTCTGTCACCCAGGCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55793546</th>\n",
       "      <td>2</td>\n",
       "      <td>230397595</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs1246211497</td>\n",
       "      <td>[T]</td>\n",
       "      <td>230397532</td>\n",
       "      <td>230397659</td>\n",
       "      <td>2:230397532-230397659</td>\n",
       "      <td>CTGCTTTGGGTTCTGCTGAAATCATGGATGAGTTCTTTCTTTAAAT...</td>\n",
       "      <td>[CTGCTTTGGGTTCTGCTGAAATCATGGATGAGTTCTTTCTTTAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15278237</th>\n",
       "      <td>2</td>\n",
       "      <td>58860529</td>\n",
       "      <td>G</td>\n",
       "      <td>A,T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>rs768583468</td>\n",
       "      <td>[A, T]</td>\n",
       "      <td>58860466</td>\n",
       "      <td>58860593</td>\n",
       "      <td>2:58860466-58860593</td>\n",
       "      <td>ATTAATATTTACCCTGTATATTTTCACAGGACCATTATATTGATCA...</td>\n",
       "      <td>[ATTAATATTTACCCTGTATATTTTCACAGGACCATTATATTGATC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr        pos ref  alt  tsa            id alt_list      start  \\\n",
       "2641070    2   10483775   C  A,T  SNV  rs1329558856   [A, T]   10483712   \n",
       "51700147   2  213255429   G    A  SNV  rs1449218878      [A]  213255366   \n",
       "27453276   2  109750792   C    T  SNV  rs1438538645      [T]  109750729   \n",
       "55793546   2  230397595   C    T  SNV  rs1246211497      [T]  230397532   \n",
       "15278237   2   58860529   G  A,T  SNV   rs768583468   [A, T]   58860466   \n",
       "\n",
       "                end                    bed  \\\n",
       "2641070    10483839    2:10483712-10483839   \n",
       "51700147  213255493  2:213255366-213255493   \n",
       "27453276  109750856  2:109750729-109750856   \n",
       "55793546  230397659  2:230397532-230397659   \n",
       "15278237   58860593    2:58860466-58860593   \n",
       "\n",
       "                                                    ref_seq  \\\n",
       "2641070   TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTGT...   \n",
       "51700147  AGTATTATTTCTATTCTTTACTTGAGAATCCAGTTTTGTAGGACTT...   \n",
       "27453276  ATTTTTTTTTTTCGAGTCAGAGTCTTGTTCTGTCACCCAGGCTGGA...   \n",
       "55793546  CTGCTTTGGGTTCTGCTGAAATCATGGATGAGTTCTTTCTTTAAAT...   \n",
       "15278237  ATTAATATTTACCCTGTATATTTTCACAGGACCATTATATTGATCA...   \n",
       "\n",
       "                                                    alt_seq  \n",
       "2641070   [TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTG...  \n",
       "51700147  [AGTATTATTTCTATTCTTTACTTGAGAATCCAGTTTTGTAGGACT...  \n",
       "27453276  [ATTTTTTTTTTTCGAGTCAGAGTCTTGTTCTGTCACCCAGGCTGG...  \n",
       "55793546  [CTGCTTTGGGTTCTGCTGAAATCATGGATGAGTTCTTTCTTTAAA...  \n",
       "15278237  [ATTAATATTTACCCTGTATATTTTCACAGGACCATTATATTGATC...  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr2_rand_samples.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To represent the reference sequences `0`, the non-associated variants `1`, the gwas-associated variants `2`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2641070</th>\n",
       "      <td>TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51700147</th>\n",
       "      <td>AGTATTATTTCTATTCTTTACTTGAGAATCCAGTTTTGTAGGACTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27453276</th>\n",
       "      <td>ATTTTTTTTTTTCGAGTCAGAGTCTTGTTCTGTCACCCAGGCTGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55793546</th>\n",
       "      <td>CTGCTTTGGGTTCTGCTGAAATCATGGATGAGTTCTTTCTTTAAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15278237</th>\n",
       "      <td>ATTAATATTTACCCTGTATATTTTCACAGGACCATTATATTGATCA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                        seq\n",
       "2641070   TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTGT...\n",
       "51700147  AGTATTATTTCTATTCTTTACTTGAGAATCCAGTTTTGTAGGACTT...\n",
       "27453276  ATTTTTTTTTTTCGAGTCAGAGTCTTGTTCTGTCACCCAGGCTGGA...\n",
       "55793546  CTGCTTTGGGTTCTGCTGAAATCATGGATGAGTTCTTTCTTTAAAT...\n",
       "15278237  ATTAATATTTACCCTGTATATTTTCACAGGACCATTATATTGATCA..."
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr2_rand_refs = chr2_rand_samples[['ref_seq']].copy(deep=True)\n",
    "chr2_rand_refs.rename(columns={'ref_seq':'seq'}, inplace=True)\n",
    "chr2_rand_refs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2641070</th>\n",
       "      <td>TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTGT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51700147</th>\n",
       "      <td>AGTATTATTTCTATTCTTTACTTGAGAATCCAGTTTTGTAGGACTT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27453276</th>\n",
       "      <td>ATTTTTTTTTTTCGAGTCAGAGTCTTGTTCTGTCACCCAGGCTGGA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55793546</th>\n",
       "      <td>CTGCTTTGGGTTCTGCTGAAATCATGGATGAGTTCTTTCTTTAAAT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15278237</th>\n",
       "      <td>ATTAATATTTACCCTGTATATTTTCACAGGACCATTATATTGATCA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                        seq  label\n",
       "2641070   TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTGT...      0\n",
       "51700147  AGTATTATTTCTATTCTTTACTTGAGAATCCAGTTTTGTAGGACTT...      0\n",
       "27453276  ATTTTTTTTTTTCGAGTCAGAGTCTTGTTCTGTCACCCAGGCTGGA...      0\n",
       "55793546  CTGCTTTGGGTTCTGCTGAAATCATGGATGAGTTCTTTCTTTAAAT...      0\n",
       "15278237  ATTAATATTTACCCTGTATATTTTCACAGGACCATTATATTGATCA...      0"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr2_rand_refs['label'] = 0\n",
    "chr2_rand_refs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2641070</th>\n",
       "      <td>TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2641070</th>\n",
       "      <td>TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51700147</th>\n",
       "      <td>AGTATTATTTCTATTCTTTACTTGAGAATCCAGTTTTGTAGGACTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27453276</th>\n",
       "      <td>ATTTTTTTTTTTCGAGTCAGAGTCTTGTTCTGTCACCCAGGCTGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55793546</th>\n",
       "      <td>CTGCTTTGGGTTCTGCTGAAATCATGGATGAGTTCTTTCTTTAAAT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                        seq\n",
       "2641070   TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTGT...\n",
       "2641070   TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTGT...\n",
       "51700147  AGTATTATTTCTATTCTTTACTTGAGAATCCAGTTTTGTAGGACTT...\n",
       "27453276  ATTTTTTTTTTTCGAGTCAGAGTCTTGTTCTGTCACCCAGGCTGGA...\n",
       "55793546  CTGCTTTGGGTTCTGCTGAAATCATGGATGAGTTCTTTCTTTAAAT..."
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr2_rand_alt = chr2_rand_samples[['alt_seq']].copy(deep=True).explode('alt_seq')\n",
    "chr2_rand_alt.rename(columns={'alt_seq':'seq'}, inplace=True)\n",
    "chr2_rand_alt.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2641070</th>\n",
       "      <td>TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTGT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2641070</th>\n",
       "      <td>TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTGT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51700147</th>\n",
       "      <td>AGTATTATTTCTATTCTTTACTTGAGAATCCAGTTTTGTAGGACTT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27453276</th>\n",
       "      <td>ATTTTTTTTTTTCGAGTCAGAGTCTTGTTCTGTCACCCAGGCTGGA...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55793546</th>\n",
       "      <td>CTGCTTTGGGTTCTGCTGAAATCATGGATGAGTTCTTTCTTTAAAT...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                        seq  label\n",
       "2641070   TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTGT...      1\n",
       "2641070   TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTGT...      1\n",
       "51700147  AGTATTATTTCTATTCTTTACTTGAGAATCCAGTTTTGTAGGACTT...      1\n",
       "27453276  ATTTTTTTTTTTCGAGTCAGAGTCTTGTTCTGTCACCCAGGCTGGA...      1\n",
       "55793546  CTGCTTTGGGTTCTGCTGAAATCATGGATGAGTTCTTTCTTTAAAT...      1"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr2_rand_alt['label'] = 1\n",
    "chr2_rand_alt.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9074</th>\n",
       "      <td>TGCAAATAGTGTATAGAAAAAGCTCTGTTTAGAAACTGCCATAGCA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14066</th>\n",
       "      <td>CTCCACAATAGTCAAAATGAAAGAAAAATACCAAGCCTCTCTCAGC...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18575</th>\n",
       "      <td>TTCTCACACTGCCAATAAAAACATAGCTAAGACTGGGTAATTTATA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18619</th>\n",
       "      <td>GGCAAAGGAGGAGCAAAGGCACCTCTTACATGGTGGCAGGCAAGAG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21715</th>\n",
       "      <td>AGTACAGGACACACCATCGCAATAAATTAAAAAGGCAAATGCAAAT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     seq  label\n",
       "9074   TGCAAATAGTGTATAGAAAAAGCTCTGTTTAGAAACTGCCATAGCA...      0\n",
       "14066  CTCCACAATAGTCAAAATGAAAGAAAAATACCAAGCCTCTCTCAGC...      0\n",
       "18575  TTCTCACACTGCCAATAAAAACATAGCTAAGACTGGGTAATTTATA...      0\n",
       "18619  GGCAAAGGAGGAGCAAAGGCACCTCTTACATGGTGGCAGGCAAGAG...      0\n",
       "21715  AGTACAGGACACACCATCGCAATAAATTAAAAAGGCAAATGCAAAT...      0"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr2_gwas_variants = pd.read_csv(os.path.join(dataset_path, 'chr2_gwas_dataset.csv'), index_col=0)\n",
    "chr2_gwas_refs = chr2_gwas_variants[['ref_seq']].copy(deep=True)\n",
    "chr2_gwas_refs.rename(columns={'ref_seq':'seq'}, inplace=True)\n",
    "chr2_gwas_refs['label'] = 0\n",
    "chr2_gwas_refs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTGT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AGTATTATTTCTATTCTTTACTTGAGAATCCAGTTTTGTAGGACTT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ATTTTTTTTTTTCGAGTCAGAGTCTTGTTCTGTCACCCAGGCTGGA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CTGCTTTGGGTTCTGCTGAAATCATGGATGAGTTCTTTCTTTAAAT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ATTAATATTTACCCTGTATATTTTCACAGGACCATTATATTGATCA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65855</th>\n",
       "      <td>TATTGTACACGCATATGTGGGTATATGTGATGACGGACGATTCGGC...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65856</th>\n",
       "      <td>CCCCGCCCTTTGCCCCTGCAGCCCCACCCCCCACCTCAGTGAAAAA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65857</th>\n",
       "      <td>TTTAGTAGAAACGGGGTTTCACCGTGTTAGCCGGGATGGACTTGAT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65858</th>\n",
       "      <td>CTCCTTCACCTGCCCCTCCGGTGACAGGAGAGTTATGAGCTAGGTC...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65859</th>\n",
       "      <td>TGTACAGATTTGTTACATGAAAATATTGCATGAGGCTGGGGTTTGG...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65860 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     seq  label\n",
       "0      TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTGT...      0\n",
       "1      AGTATTATTTCTATTCTTTACTTGAGAATCCAGTTTTGTAGGACTT...      0\n",
       "2      ATTTTTTTTTTTCGAGTCAGAGTCTTGTTCTGTCACCCAGGCTGGA...      0\n",
       "3      CTGCTTTGGGTTCTGCTGAAATCATGGATGAGTTCTTTCTTTAAAT...      0\n",
       "4      ATTAATATTTACCCTGTATATTTTCACAGGACCATTATATTGATCA...      0\n",
       "...                                                  ...    ...\n",
       "65855  TATTGTACACGCATATGTGGGTATATGTGATGACGGACGATTCGGC...      0\n",
       "65856  CCCCGCCCTTTGCCCCTGCAGCCCCACCCCCCACCTCAGTGAAAAA...      0\n",
       "65857  TTTAGTAGAAACGGGGTTTCACCGTGTTAGCCGGGATGGACTTGAT...      0\n",
       "65858  CTCCTTCACCTGCCCCTCCGGTGACAGGAGAGTTATGAGCTAGGTC...      0\n",
       "65859  TGTACAGATTTGTTACATGAAAATATTGCATGAGGCTGGGGTTTGG...      0\n",
       "\n",
       "[65860 rows x 2 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr2_dataset = pd.concat([chr2_rand_refs, chr2_rand_alt, chr2_gwas_refs], ignore_index = True)\n",
    "chr2_dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the resulting dataframe as it is ready to get into the dataloader\n",
    "#chr2_dataset.to_csv(os.path.join(ensembl_path, 'to_dataloaders/chr2_dataset.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTGT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AGTATTATTTCTATTCTTTACTTGAGAATCCAGTTTTGTAGGACTT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ATTTTTTTTTTTCGAGTCAGAGTCTTGTTCTGTCACCCAGGCTGGA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CTGCTTTGGGTTCTGCTGAAATCATGGATGAGTTCTTTCTTTAAAT...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ATTAATATTTACCCTGTATATTTTCACAGGACCATTATATTGATCA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 seq  label\n",
       "0  TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTGT...      0\n",
       "1  AGTATTATTTCTATTCTTTACTTGAGAATCCAGTTTTGTAGGACTT...      0\n",
       "2  ATTTTTTTTTTTCGAGTCAGAGTCTTGTTCTGTCACCCAGGCTGGA...      0\n",
       "3  CTGCTTTGGGTTCTGCTGAAATCATGGATGAGTTCTTTCTTTAAAT...      0\n",
       "4  ATTAATATTTACCCTGTATATTTTCACAGGACCATTATATTGATCA...      0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr2_dataset = pd.read_csv(os.path.join(ensembl_path, 'to_dataloaders/chr2_dataset.csv'), index_col=0)\n",
    "chr2_dataset.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "label\n",
       "0    42464\n",
       "1    23396\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr2_dataset['label'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'TTCTTTTAAATGGCTACATAATTAAGTCTAAGGTGAGAATTACTGTACTTTATTCAACCCTCTCGTTGATAGGTATTCACTTTGTTCCCACAGTTTTCCAGGGCTACAATCCTTGAATATCAAGCCTT'"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr2_dataset['seq'].iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32395 21232\n"
     ]
    }
   ],
   "source": [
    "print(len(chr2_rand_alt), len(chr2_rand_refs))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Creating the DataLoader\n",
    "To store our dataset the PyTorch library provides two data primitives: `torch.utils.data.DataLoader` and `torch.utils.data.Dataset`. `Dataset` stores the samples and their corresponding labels, while `DataLoader` wraps an iterable around the `Dataset` and retrieves them easily.\n",
    "\n",
    "A custom `Dataset` class must contain at least three functions `__init__`, `__getitem__`, and `__len__`. These classes inherit from `Dataset` in `torch.utils.data.Dataset`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create the Dataset class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from transformers import AutoTokenizer, DataCollatorWithPadding\n",
    "from datasets import load_dataset\n",
    "from datasets import Dataset\n",
    "tokenizer = AutoTokenizer.from_pretrained(\"zhihan1996/DNABERT-2-117M\", trust_remote_code=True)\n",
    "data_collator = DataCollatorWithPadding(tokenizer=tokenizer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Este hay que dejarlo por si acaso para referencia\n",
    "class VariantDataset(Dataset):\n",
    "    def __init__(self, dataset_path, transform=None, target_transform=None):\n",
    "        self.dataset = pd.read_csv(dataset_path)\n",
    "        self.transform = transform\n",
    "        self.target_transform = target_transform\n",
    "\n",
    "    def __len__(self):\n",
    "        return len(self.dataset)\n",
    "\n",
    "    def __getitem__(self, idx):\n",
    "        sequence = self.dataset['seq'].iloc[idx]\n",
    "        label = self.dataset['label'].iloc[idx]\n",
    "        return sequence, label\n",
    "    \n",
    "#test_dataset = VariantDataset(os.path.join(ensembl_path, 'to_dataloaders/chr2_dataset.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DatasetDict({\n",
       "    train: Dataset({\n",
       "        features: ['seq', 'label', '__index_level_0__'],\n",
       "        num_rows: 46102\n",
       "    })\n",
       "    test: Dataset({\n",
       "        features: ['seq', 'label', '__index_level_0__'],\n",
       "        num_rows: 19758\n",
       "    })\n",
       "})"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Using the HuggingFace framework to create the Datasets and DataLoaders\n",
    "hf_dataset = Dataset.from_pandas(chr2_dataset).train_test_split(test_size=0.3)\n",
    "hf_dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "32cf9a1b2b98460e948b501fd86ddc5d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Map:   0%|          | 0/46102 [00:00<?, ? examples/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d216b864e3ee4bcfa99661f52a855439",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Map:   0%|          | 0/19758 [00:00<?, ? examples/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "DatasetDict({\n",
       "    train: Dataset({\n",
       "        features: ['seq', 'label', '__index_level_0__', 'input_ids', 'token_type_ids', 'attention_mask'],\n",
       "        num_rows: 46102\n",
       "    })\n",
       "    test: Dataset({\n",
       "        features: ['seq', 'label', '__index_level_0__', 'input_ids', 'token_type_ids', 'attention_mask'],\n",
       "        num_rows: 19758\n",
       "    })\n",
       "})"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def tokenize_function(example):\n",
    "    return tokenizer(example[\"seq\"], truncation=True)\n",
    "\n",
    "tokenized_datasets = hf_dataset.map(tokenize_function, batched=True)\n",
    "tokenized_datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'train': ['label', 'input_ids', 'token_type_ids', 'attention_mask'],\n",
       " 'test': ['label', 'input_ids', 'token_type_ids', 'attention_mask']}"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tokenized_datasets = tokenized_datasets.remove_columns([\"seq\", \"__index_level_0__\"])\n",
    "tokenized_datasets.set_format(\"torch\")\n",
    "tokenized_datasets.column_names"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create the DataLoader\n",
    "The `DataLoader` will wrap an iterable around the `Dataset` object to retrive samples in an efficient manner in order to train the neural network."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_set = DataLoader(tokenized_datasets[\"train\"], batch_size=8, shuffle=True, collate_fn=data_collator)\n",
    "test_set = DataLoader(tokenized_datasets[\"test\"], batch_size=8, shuffle=True, collate_fn=data_collator)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'input_ids': torch.Size([8, 29]),\n",
       " 'token_type_ids': torch.Size([8, 29]),\n",
       " 'attention_mask': torch.Size([8, 29]),\n",
       " 'labels': torch.Size([8])}"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for batch in train_set:\n",
    "    break\n",
    "{k: v.shape for k, v in batch.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'input_ids': tensor([[   1,   10,   35,  438,   47,  147,   54,  123,   33, 1079,   64, 2467,\n",
      "         3107,  105, 3235,   83, 1022,   86,  279, 1104,  141, 3168,  261,  999,\n",
      "          213,   92,   22,    2,    3,    3],\n",
      "        [   1,    5,   45,  840,  798,   35,   89,  427,  274,   67,  373,   32,\n",
      "          132,   82, 1410, 3415,  785,   92,   73,   48,  101, 1367,  119,   76,\n",
      "          531,   32,  215,  769,    2,    3],\n",
      "        [   1,    9, 3512,  349,  242,   85, 2574, 2058,  138,   92,  109,   17,\n",
      "          590,  135,  742,   81,  202,  174,  823,  317,   70,   62,  485, 1908,\n",
      "           82,  727,    6,    2,    3,    3],\n",
      "        [   1,   74,  189, 1293,  120,   17, 1453,   65,   95,  339,   43, 1005,\n",
      "          256,   78, 3508,  233,   72,  106,  121,   52,   65,  924,  103, 1828,\n",
      "         1937,  420,  615,    2,    3,    3],\n",
      "        [   1,   11,   68,  222,   88,  491,  474,   61,  710,   61, 1326, 3606,\n",
      "          482,   66, 2406,  722,  100,  921,   45, 3382,  503,   29,  347,  500,\n",
      "          324,   70,   56,    2,    3,    3],\n",
      "        [   1,    9,  226,   25,  237,   28, 2096,  160,  101,  328,   85, 2146,\n",
      "           20, 1270,  696,   44,  418,  505,  309,   32,  175,   27,   19,  395,\n",
      "          184,   31,  168,   51,   15,    2],\n",
      "        [   1,  713,  103,  277,   73,   33,  263,  369,  274,   83,   50, 1860,\n",
      "           70, 1542,  228,   67,  745,  430,   35,  320,  310,   58,   98,  462,\n",
      "           27,  215,   12, 1049,    2,    3],\n",
      "        [   1,  204,  223,  105,  474,  127,   25,  679,   61,   51,  212,  172,\n",
      "         2205,   49,  328,  713,  535,   57,  870,   24,  112,   87,   37,  132,\n",
      "          271,   75,  293,  243,    2,    3]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
      "         0, 0, 0, 0, 0, 0],\n",
      "        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
      "         0, 0, 0, 0, 0, 0],\n",
      "        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
      "         0, 0, 0, 0, 0, 0],\n",
      "        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
      "         0, 0, 0, 0, 0, 0],\n",
      "        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
      "         0, 0, 0, 0, 0, 0],\n",
      "        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
      "         0, 0, 0, 0, 0, 0],\n",
      "        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
      "         0, 0, 0, 0, 0, 0],\n",
      "        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
      "         0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n",
      "         1, 1, 1, 1, 0, 0],\n",
      "        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n",
      "         1, 1, 1, 1, 1, 0],\n",
      "        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n",
      "         1, 1, 1, 1, 0, 0],\n",
      "        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n",
      "         1, 1, 1, 1, 0, 0],\n",
      "        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n",
      "         1, 1, 1, 1, 0, 0],\n",
      "        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n",
      "         1, 1, 1, 1, 1, 1],\n",
      "        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n",
      "         1, 1, 1, 1, 1, 0],\n",
      "        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n",
      "         1, 1, 1, 1, 1, 0]]), 'labels': tensor([1, 1, 0, 1, 0, 0, 1, 1])}\n"
     ]
    }
   ],
   "source": [
    "first_batch = next(iter(train_set))\n",
    "print(first_batch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "DNABERT_2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
